File: iossdkdemo/SDKDemo/APIControllers/GeofencingViewController.m
//
//  GeofencingViewController.m
//  SDKDemoApp
//
//  Created by OnePoint Global on 06/10/16.
//  Copyright © 2016 opg. All rights reserved.
//

#import "GeofencingViewController.h"

@interface GeofencingViewController ()<OPGGeoFencingDelegate,UITableViewDelegate,UITableViewDataSource> {
    OPGGeoFencing* geo;
    NSMutableArray* arrayLocations;
    NSArray *enabledSurveys;
    NSArray* surveyList;
}
@end

@implementation GeofencingViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    NSError* error;
    OPGSDK* sdk = [OPGSDK new];
    OPGAuthenticate* authenticate = [sdk authenticate:@"USERNAME" password:@"PASSWORD" error:&error];
    
    if ([authenticate.isSuccess  isEqual: @1]) {
        NSLog(@"Authentication Success");
    } else {
        NSLog(@"%@", authenticate.statusMessage);
    }
    surveyList = [sdk getUserSurveyList:&error];
    geo = [[OPGGeoFencing alloc] initWithGeoFencingSurveys:surveyList];
    geo.geoFencingDelegate = self;
    
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)switchTapped:(id)sender {
    UISwitch* customSwitch = (UISwitch*)sender;
    if (customSwitch.isOn) {
        [geo startGeoFencing];
        
    } else if (!customSwitch.isOn ) {
        [geo stopGeoFencing];
        
    }
}

- (void)showAlert{
    if ([arrayLocations count] == 0) {
        return;
    }
    UIAlertController * alert = [UIAlertController
                                 alertControllerWithTitle:@"OPGSDK"
                                 message:[NSString stringWithFormat:@"Congratulations! You're at location %@ and a survey available for u at this location",[[arrayLocations firstObject] address]]
                                 preferredStyle:UIAlertControllerStyleAlert];
    
    UIAlertAction* okButton = [UIAlertAction
                               actionWithTitle:@"OK"
                               style:UIAlertActionStyleDefault
                               handler:^(UIAlertAction * action) {
                                   [arrayLocations removeObjectAtIndex:0];
                                   [self performSelector:@selector(showAlert) withObject:nil afterDelay:0.0];
                                   
                               }];
    UIAlertAction* takeSurveyButton = [UIAlertAction
                                       actionWithTitle:@"TakeSurvey"
                                       style:UIAlertActionStyleDefault
                                       handler:^(UIAlertAction * action) {
                                       }];
    
    [alert addAction:okButton];
    [alert addAction:takeSurveyButton];
    [self presentViewController:alert animated:YES completion:nil];
    
}

-(void)didEnterSurveyRegion:(NSArray *)locations {
    NSLog(@"Entered locations %@",locations);
    arrayLocations = [NSMutableArray arrayWithArray:locations];
    
    NSMutableArray* surveys = [NSMutableArray new];
    
    if ([locations count] > 0) {
        [self showAlert];
    }
    
    for (OPGGeoFencingModel* model in locations) {
        [surveys addObject:model.surveyName];
    }
    
    NSOrderedSet *orderedSet = [NSOrderedSet orderedSetWithArray:surveys];
    enabledSurveys = [orderedSet array];
    
}

-(void)didExitSurveyRegion:(NSArray *)locations {
    NSLog(@"Exited locations %@",locations);
    
}

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return  surveyList.count;
}

-(CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 60.0;
}

-(UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell =  [tableView dequeueReusableCellWithIdentifier:@"customCell"];
    if (!cell) {
        cell = [[UITableViewCell alloc]initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"customCell"];
    }
    OPGSurvey* survey = [surveyList objectAtIndex:indexPath.row];
    cell.textLabel.text = survey.surveyName;
    return cell;
}


@end

File: iossdkdemo/SDKDemo-Swift/APIControllers-Swift/GeofencingViewController.swift
//
//  GeofencingViewController.swift
//  SDKDemoApp
//
//  Created by OnePoint Global on 06/10/16.
//  Copyright © 2016 opg. All rights reserved.
//

import UIKit

class GeofencingViewController: UIViewController,OPGGeoFencingDelegate
{
    
    var geo : OPGGeoFencing!
    @IBOutlet weak var switchControl : UISwitch!
    var arrayLocations : NSMutableArray = []
    var surveyList : Array<Any> = []
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        let sdk = OPGSDK()
        var auth = OPGAuthenticate()
        
        do {
            auth =  try sdk.authenticate("USERNAME", password: "PASSWORD")
        } catch  {
            print("Authentication Error")
        }
        
        if (auth.isSuccess.intValue == 1 ) {
            print("Authentication Sucessful")
        }
        do {
            surveyList = try sdk.getUserSurveyList() as Array<Any>       // Getting list of Surveys assigned to user
        }
        catch{
            print("User Survey not available/working")         /* @"Error Occured. Contact Support!" */
            
        }
        self.geo = OPGGeoFencing.init(geoFencingSurveys: surveyList )
        self.geo.geoFencingDelegate = self
        
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    @IBAction func switchTapped(_ sender: AnyObject) {
        let custom : UISwitch = sender as! UISwitch
        if custom.isOn {
            self.geo.startGeoFencing()
        } else {
            self.geo.stopGeoFencing()
        }
    }
    
    func didEnterSurveyRegion(_ locations: [Any]!) {
        self.arrayLocations = NSMutableArray.init(array: locations)
        
        if self.arrayLocations.count > 0 {
            self.showAlert()
        }
    }
    
    func didExitSurveyRegion(_ locations: [Any]!) {
        
    }
    
    func showAlert() {
        if self.arrayLocations.count == 0 {
            return
        }
        let model : OPGGeoFencingModel = self.arrayLocations.firstObject as! OPGGeoFencingModel
        let alert : UIAlertController = UIAlertController.init(title: "OPGSDK", message: "Congratulations! You're at location \(model.address) and a survey available for u at this location", preferredStyle: UIAlertControllerStyle.alert)
        alert.addAction(UIAlertAction(title: "OK", style: UIAlertActionStyle.default, handler: {(action:UIAlertAction) in
            self.arrayLocations.removeObject(at: 0)
            self.showAlert()
        }))
        alert.addAction(UIAlertAction.init(title: "Cancel", style: UIAlertActionStyle.cancel, handler: nil))
        present(alert, animated: true, completion: nil);
    }

}
